File: app/src/main/java/com/amazon/tv/leanbacklauncher/settings/LegacyFileListFragment.java
package com.amazon.tv.leanbacklauncher.settings;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.res.ResourcesCompat;
import androidx.leanback.app.GuidedStepSupportFragment;
import androidx.leanback.widget.GuidanceStylist.Guidance;
import androidx.leanback.widget.GuidedAction;
import androidx.leanback.widget.GuidedAction.Builder;

import com.amazon.tv.leanbacklauncher.MainActivity;
import com.amazon.tv.leanbacklauncher.R;

import java.io.File;
import java.util.List;

public class LegacyFileListFragment extends GuidedStepSupportFragment {

    private static String TAG = "LegacyFileListFragment";
    /* Action ID definition */
    private static final int ACTION_SELECT = 1;
    private static final int ACTION_BACK = 2;


    @NonNull
    public Guidance onCreateGuidance(Bundle savedInstanceState) {
        Activity activity = getActivity();
        String desc = getWallpaper(activity);
        return new Guidance(
                getString(R.string.select_wallpaper_action_title), // title
                desc, // description
                getString(R.string.settings_dialog_title), // breadcrumb (parent)
                ResourcesCompat.getDrawable(getResources(), R.drawable.ic_settings_home, null) // icon
        );
    }

    public void onCreateActions(@NonNull List<GuidedAction> actions, Bundle savedInstanceState) {
        Activity activity = getActivity();
        if (ContextCompat.checkSelfPermission(activity, android.Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
            Log.v(TAG, "READ_EXTERNAL_STORAGE permission is granted");
        } else {
            Log.v(TAG, "READ_EXTERNAL_STORAGE permission not granted");
            makeRequest();
        }
        actions.add(new Builder(
                getActivity())
                .id(ACTION_SELECT)
                .title(String.format(getString(R.string.select), Environment.getExternalStorageDirectory()))
                .description(null)
                .build()
        );
        actions.add(new Builder(
                getActivity())
                .id(ACTION_BACK)
                .title(R.string.goback)
                .description(null)
                .build()
        );
    }

    protected void makeRequest() {
        Activity activity = getActivity();
        ActivityCompat.requestPermissions(activity,
                new String[]{android.Manifest.permission.READ_EXTERNAL_STORAGE},
                500);
    }

    public void onGuidedActionClicked(GuidedAction action) {
        Activity activity = getActivity();
        switch ((int) action.getId()) {
            case ACTION_SELECT:
                File file = new File(Environment.getExternalStorageDirectory(), "background.jpg");
                if (file.canRead())
                    setWallpaper(activity, Environment.getExternalStorageDirectory() + "/background.jpg");
                else
                    Toast.makeText(activity, activity.getString(R.string.file_no_access), Toast.LENGTH_LONG).show();
                getFragmentManager().popBackStack();
                break;
            case ACTION_BACK:
                getFragmentManager().popBackStack();
                break;
            default:
                break;
        }
    }

    public boolean setWallpaper(Context context, String image) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        if (!image.isEmpty())
            pref.edit().putString("wallpaper_image", image).apply();
        // refresh home
        Activity activity = getActivity();
        Intent Broadcast = new Intent(MainActivity.class.getName()); // ACTION
        Broadcast.putExtra("RefreshHome", true);
        activity.sendBroadcast(Broadcast);
        return true;
    }

    private String getWallpaper(Context context) {
        SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
        String image = pref.getString("wallpaper_image", "");
        if (!image.isEmpty()) {
            return image;
        } else {
            File file = new File(context.getFilesDir(), "background.jpg");
            if (file.canRead()) {
                return file.toString();
            }
            return null;
        }
    }
}

File: app/src/main/java/com/amazon/tv/leanbacklauncher/settings/LegacyFileListFragment.kt
package com.amazon.tv.leanbacklauncher.settings

import android.content.Context
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.Toast
import androidx.core.content.res.ResourcesCompat
import androidx.leanback.app.GuidedStepSupportFragment
import androidx.leanback.widget.GuidanceStylist.Guidance
import androidx.leanback.widget.GuidedAction
import androidx.preference.PreferenceManager
import com.amazon.tv.leanbacklauncher.BuildConfig
import com.amazon.tv.leanbacklauncher.R
import com.amazon.tv.leanbacklauncher.util.Util.refreshHome
import java.io.File

class LegacyFileListFragment : GuidedStepSupportFragment() {

    companion object {
        private val TAG by lazy {
            if (BuildConfig.DEBUG) ("[*]" + LegacyFileListFragment::class.java.simpleName).take(
                21
            ) else LegacyFileListFragment::class.java.simpleName
        }
        private var rootPath: String? = null
        private var dirName: String? = null

        /* Action ID definition */
        private const val ACTION_SELECT = 1
        private const val ACTION_BACK = 2
        private const val ACTION_DIR = 3
    }

    override fun onCreateGuidance(savedInstanceState: Bundle?): Guidance {
        return Guidance(
            getString(R.string.select_wallpaper_action_title),  // title
            getWallpaperDesc(requireContext()),  // description
            getString(R.string.settings_dialog_title),  // breadcrumb (parent)
            ResourcesCompat.getDrawable(resources, R.drawable.ic_settings_home, null) // icon
        )
    }

    override fun onResume() {
        super.onResume()
        val fm = fragmentManager
        if (rootPath.isNullOrEmpty() || fm?.backStackEntryCount == 3) // 3 on root dir
            rootPath = Environment.getExternalStorageDirectory().absolutePath
        val ctx = requireContext()
        if (BuildConfig.DEBUG) Log.d(TAG, "onResume() rootPath: $rootPath")

        var dir: File? = null
        var subdirs: ArrayList<File> = ArrayList()
        var dimages: ArrayList<File> = ArrayList()
        rootPath?.let { dir = File(it) }
        dir?.let {
            subdirs = dirReader(it)
            dimages = imageReader(it)
        }
        val actions = ArrayList<GuidedAction>()
        // back
        actions.add(
            GuidedAction.Builder(ctx)
                .id(ACTION_BACK.toLong())
                .title(R.string.goback)
                .description(null)
                .build()
        )
        // directories
        if (subdirs.size > 0)
            subdirs.forEach {
                actions.add(
                    GuidedAction.Builder(ctx)
                        .id(ACTION_DIR.toLong())
                        .title(it.name)
                        .description(null)
                        .build()
                )
            }
        // images
        if (dimages.size > 0)
            dimages.forEach {
                actions.add(
                    GuidedAction.Builder(ctx)
                        .id(ACTION_SELECT.toLong())
                        .title(it.name)
                        .description(null)
                        .build()
                )
            }

        setActions(actions)
    }

    override fun onGuidedActionClicked(action: GuidedAction) {
        val ctx = requireContext()
        val fm = fragmentManager
        when (action.id.toInt()) {
            ACTION_DIR -> {
                rootPath = File(rootPath, action.title.toString()).absolutePath
                dirName = action.title.toString()
                if (fm != null) {
                    add(fm, LegacyFileListFragment())
                }
            }
            ACTION_SELECT -> {
                val name = action.title
                val file = File(rootPath, name.toString())
                if (file.canRead())
                    setWallpaper(ctx, file.path.toString())
                else
                    Toast.makeText(ctx, ctx.getString(R.string.file_no_access), Toast.LENGTH_LONG)
                        .show()
                dirName?.let { rootPath = rootPath?.removeSuffix(it) }
                fm?.popBackStack()
            }
            ACTION_BACK -> {
                dirName?.let { rootPath = rootPath?.removeSuffix(it) }
                fm?.popBackStack()
            }
        }
        //if (BuildConfig.DEBUG) Log.d(TAG, "onGuidedActionClicked() rootPath: $rootPath, backstack: ${fm?.backStackEntryCount}")
    }

    fun imageReader(root: File): ArrayList<File> {
        val imageList: ArrayList<File> = ArrayList()
        val images = root.listFiles { file ->
            (!file.isDirectory && file.name.endsWith(".jpeg") || file.name.endsWith(".jpg") || file.name.endsWith(
                ".png"
            ))
        }
        if (!images.isNullOrEmpty()) {
            for (image in images) {
                // File absolute path
                if (BuildConfig.DEBUG) Log.d(TAG, "image path ${image.absolutePath}")
                // File Name
                if (BuildConfig.DEBUG) Log.d(TAG, "image name ${image.name}")
                imageList.add(image.absoluteFile)
            }
            if (BuildConfig.DEBUG) Log.w(TAG, "fileList size ${imageList.size}")
        }
        return imageList
    }

    fun dirReader(root: File): ArrayList<File> {
        val dirList: ArrayList<File> = ArrayList()
        val dirs = root.listFiles { file ->
            file.isDirectory && !file.name.startsWith(".")
        }
        if (!dirs.isNullOrEmpty())
            for (dir in dirs) {
                dirList.add(dir.absoluteFile)
            }
        return dirList
    }

    private fun setWallpaper(context: Context?, image: String): Boolean {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        if (image.isNotEmpty()) pref.edit().putString("wallpaper_image", image).apply()
        // refresh home broadcast
        val activity = requireActivity()
        refreshHome(activity)
        return true
    }

    private fun getWallpaperDesc(context: Context): String {
        val pref = PreferenceManager.getDefaultSharedPreferences(context)
        val imagePath = pref.getString("wallpaper_image", "")
        return if (imagePath!!.isNotBlank()) {
            imagePath
        } else {
            val file = File(context.getExternalFilesDir(null), "background.jpg")
            if (file.canRead()) {
                file.toString()
            }
            getString(R.string.wallpaper_choose)
        }
    }
}